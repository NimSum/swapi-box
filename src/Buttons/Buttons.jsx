import React from 'react';

export const PeopleBtn= () => {
  return (
    <button 
      type="button"
      className="category-button">
      People
    </button>
  )
}

export const PlanetsBtn = () => {
  return (
    <button 
      type="button"
      className="category-button">
      Planets
    </button>
  )
}

export const VehiclesBtn = () => {
  return (
    <button 
      type="button"
      className="category-button">
      Vehicles
    </button>
  )
}

export const ViewFavoritesBtn = ({ total }) => {
  return (
    <button 
      type="button"
      className="people-button">
      Favorites 
      <span className="fav-total">{ total }</span>
    </button>
  )
}

export const HomeBtn = () => {
  return (
    <button 
      type="button"
      className="home-button">
      Home
    </button>
  )
}