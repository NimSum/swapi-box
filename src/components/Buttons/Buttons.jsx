import React, { Component } from 'react';
import PropTypes from 'prop-types';

export const PeopleBtn= ({ setActive, activeBtn }) => {
  return (
    <button 
      type="button"
      className={`${activeBtn === "people" 
        ? "active category-btn"
        : "category-btn"}`}
      name="people"
      onClick={ setActive }>
      <img name="people" alt='card background' src={require('../../images/icons/vader.svg')} />
      People
    </button>
  )
}

export const PlanetsBtn = ({ setActive, activeBtn }) => {
  return (
    <button 
      type="button"
      className={`${activeBtn === "planets" 
        ? "active category-btn"
        : "category-btn"}`}
      name="planets"
      onClick={ setActive }>
      <img name="planets" alt='card background' src={require('../../images/icons/death-star.svg')} />
      Planets
    </button>
  )
}

export const VehiclesBtn = ({ setActive, activeBtn }) => {
  return (
    <button 
      type="button"
      className={`${activeBtn === "vehicles" 
        ? "active category-btn"
        : "category-btn"}`}
      name="vehicles"
      onClick={ setActive }>
      <img name="vehicles" alt='card background' src={require('../../images/icons/tie-fighter.svg')} />
      Vehicles
    </button>
  )
}

export const ViewFavoritesBtn = ({ activeBtn, setActive, favoriteCount }) =>  (
    <button 
      type="button"
      className={`${ activeBtn === "favorites" 
        ? "active view-fav-btn"
        : "view-fav-btn"}`}
      onClick={ setActive }
      name="favorites"
      disabled={ !favoriteCount > 0 ? true : false }>
      { !favoriteCount > 0 ? 'No favorites' : 'Favorites' }
      { favoriteCount > 0 && (<span id="favorites" className="fav-total">{ favoriteCount }</span>) }
    </button>
  )

export const HomeBtn = ({ setActive, activeBtn }) => {
  return (
    <button 
      type="button"
      className={ activeBtn === 'home' ? 'active home-btn' : 'home-btn' }
      name="home"
      onClick={ setActive } >
      Home
    </button>
  )
}

export class FavoriteCardBtn extends Component {
  constructor() {
    super();
    this.state = {
      active: false
    }
  }

  componentDidMount() {
    const storedCopy = JSON.parse(localStorage.getItem('favorites')) || [];
    storedCopy.forEach(card => card.name === this.props.card.name && this.setState({ active: true }));
  }

  saveCard = (storedCopy) => {
    const updatedCopy = storedCopy.concat([ this.props.card ]);
    localStorage.setItem('favorites', JSON.stringify(updatedCopy));
    this.props.updateFavoriteCount()
  }

  deleteCard = (storedCopy) => {
    const updatedCopy = storedCopy.filter(card => card.url !== this.props.card.url);
    localStorage.setItem('favorites', JSON.stringify(updatedCopy));
    this.props.updateFavoriteCount()
  }

  toggleFavorite = () => {
    const storedCopy = JSON.parse(localStorage.getItem('favorites')) || [];
    this.setState({ active: !this.state.active}, () => {
      this.state.active 
        ? this.saveCard(storedCopy)
        : this.deleteCard(storedCopy)
    }) 
  }

  render() {
    return (
      <button 
        className={this.state.active ? 'active-fav favorite-card-btn' : 'favorite-card-btn'}
        name="favorite-card"
        title="Add to Favorites"
        type="button"
        onClick={this.toggleFavorite}>
        <img alt='card background' src={require('../../images/icons/empire.png')} />
        <img alt='card background' src={require(`../../images/icons/${this.state.active ? 'active-star.svg' : 'inactive-star.svg'}`)} />
        <img alt='card background' src={require('../../images/icons/rebel-alliance.png')} />
      </button>
    )
  } 
}

PeopleBtn.propTypes = {
  setActive: PropTypes.func,
  activeBtn: PropTypes.string
}
PlanetsBtn.propTypes = {
  setActive: PropTypes.func,
  activeBtn: PropTypes.string
}
VehiclesBtn.propTypes = {
  setActive: PropTypes.func,
  activeBtn: PropTypes.string
}
ViewFavoritesBtn.propTypes = {
  activeBtn: PropTypes.string, 
  setActive: PropTypes.func, 
  favoriteCount: PropTypes.number 
}
HomeBtn.propTypes = {
  activeBtn: PropTypes.string,
  setActive: PropTypes.func
}
FavoriteCardBtn.propTypes = {
  card: PropTypes.object
}