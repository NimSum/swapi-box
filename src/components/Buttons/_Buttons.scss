button.active {
  background-color: $active-btn;
}