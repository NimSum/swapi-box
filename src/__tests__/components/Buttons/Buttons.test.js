import React from 'react';
import { shallow } from 'enzyme';
import { PlanetsBtn, PeopleBtn, VehiclesBtn, ViewFavoritesBtn, FavoriteCardBtn, HomeBtn, ShowMoreLessBtns } from '../../../components/Buttons/Buttons';

describe('Buttons', () => {

  const mockSetActive = jest.fn();
  afterEach(() => mockSetActive.mockClear());
  
  it('Home button: render crawl on click', () => {
    const mockActiveBtn = 'home';
    const wrapper = shallow(
      < HomeBtn 
        setActive={ mockSetActive } 
        activeBtn={ mockActiveBtn } />
    )
    wrapper.find('button').simulate('click');
    expect(mockSetActive).toHaveBeenCalledTimes(1);
  })

  it('Planet Button: trigger setActive button on click', () => {
    const mockActiveBtn = 'planets';
    const wrapper = shallow( 
      < PlanetsBtn 
        setActive={ mockSetActive }
        activeBtn={ mockActiveBtn } /> 
    );
    wrapper.find('button').simulate('click');
    expect(mockSetActive).toHaveBeenCalledTimes(1);
  })

  it('People Button: trigger setActive button on click', () => {
    const mockActiveBtn = 'people';
    const wrapper = shallow( 
      < PeopleBtn 
        setActive={ mockSetActive }
        activeBtn={ mockActiveBtn } /> 
    );
    wrapper.find('button').simulate('click');
    expect(mockSetActive).toHaveBeenCalledTimes(1);
  })

  it('Vehicle Button: trigger setActive button on click', () => {
    const mockActiveBtn = 'vehicles';
    const wrapper = shallow( 
      < VehiclesBtn 
        setActive={ mockSetActive }
        activeBtn={ mockActiveBtn } /> 
    );
    wrapper.find('button').simulate('click');
    expect(mockSetActive).toHaveBeenCalledTimes(1);
  })

  describe('ViewFavorites Button:', () => {
    const mockActiveBtn = 'favorites';
    let mockFavoriteCount = 1;

    let wrapper = shallow( 
      < ViewFavoritesBtn 
        setActive={ mockSetActive }
        activeBtn={ mockActiveBtn }
        favoriteCount={ mockFavoriteCount }/>)

    it('Should match snapshot taken', () => {
      expect(wrapper).toMatchSnapshot();
    })

    it('Should trigger change category on click', () => {
      wrapper.find('button').simulate('click', () => {
        expect(mockSetActive).toHaveBeenCalledTimes(1);
      })
    })

    it('Should be disabled with favorite count set to 0', () => {
      mockFavoriteCount = 0;
      wrapper = shallow( 
        < ViewFavoritesBtn 
          setActive={ mockSetActive }
          activeBtn={ mockActiveBtn }
          favoriteCount={ mockFavoriteCount }/>)
      expect(wrapper).toMatchSnapshot();
    })
  })

  describe('Favorite Card Button', () => {
    const mockCard = { name: 'Luke Skywalker', id: 1 }
    const mockUpdateFavCount = jest.fn();
    const mockStoredCard = [ { name: 'Luke Skywalker', id: 1 } ];
    let saveCardSpy;
    let deleteCardSpy;
    let toggleFavoriteSpy;
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(
        < FavoriteCardBtn 
          card={ mockCard }
          updateFavoriteCount={ mockUpdateFavCount } />
      )
      localStorage.setItem('favorites', JSON.stringify(mockStoredCard))
      saveCardSpy = jest.spyOn(wrapper.instance(), 'saveCard');
      deleteCardSpy = jest.spyOn(wrapper.instance(), 'deleteCard');
      toggleFavoriteSpy = jest.spyOn(wrapper.instance(), 'toggleFavorite');
    })

    it('Should be disabled/enabled if active state is true/false', () => {
      expect(wrapper).toMatchSnapshot();
      wrapper.setState({ active: true })
      expect(wrapper).toMatchSnapshot();
    })

    it('Should set button active if already favorited', () => {
      wrapper.instance().componentDidMount();
      expect(wrapper.state().active).toEqual(true);
    })

    it('Should toggle favorite button', () => {
      
      wrapper.instance().toggleFavorite();
      expect(wrapper.state().active).toEqual(false);
      expect(deleteCardSpy).toHaveBeenCalledTimes(1);

      wrapper.instance().toggleFavorite();
      expect(wrapper.state().active).toEqual(true);
      expect(saveCardSpy).toHaveBeenCalledTimes(1);
    })

    it('Should togge favorite on button click', () => {
      wrapper.find('button').simulate('click', () => {
        expect(toggleFavoriteSpy).toHaveBeenCalledTimes(1);
      })
    })

  })

  describe('Show more less buttons', () => {

    const mockShowMore = jest.fn();
    const mockShowLess = jest.fn();
    const mockCurrentPage = 1;

    let wrapper;
    beforeEach(() => {
      wrapper = shallow(
        < ShowMoreLessBtns
          showMore={ mockShowMore } 
          showLess={ mockShowLess }
          mockCurrentPage= { mockCurrentPage }
          />
      )
    })

    it('Should match component snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    })

    it('Should match snapshot of page change', () => {
      const mockPageChange = 2;
      wrapper = shallow( < ShowMoreLessBtns currentPage={ mockPageChange } />)
      expect(wrapper).toMatchSnapshot();
    })

    it('Should trigger showmore method on click', () => {
      wrapper.find('button:first-child').simulate('click', () => {
        expect(mockShowMore).toHaveBeenCalledTimes(1);
      })
    })

    it('Should trigger showmore method on click', () => {
      const mockPageChange = 2;
      wrapper = shallow( 
        < ShowMoreLessBtns 
        currentPage={ mockPageChange } 
        showLes= { mockShowLess }/>
      )

      wrapper.find('button:last-child').simulate('click', () => {
        expect(mockShowLess).toHaveBeenCalledTimes(1);
      })
    })
    
  })

})
