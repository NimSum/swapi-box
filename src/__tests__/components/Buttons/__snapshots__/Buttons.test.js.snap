// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttons Favorite Card Button Should be disabled/enabled if active state is true/false 1`] = `
<button
  className="favorite-card-btn"
  name="favorite-card"
  onClick={[Function]}
  title="Add to Favorites"
  type="button"
>
  <img
    alt="card background"
    src="empire.png"
  />
  <img
    alt="card background"
    src={
      Object {
        "ReactComponent": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "__esModule": true,
        "default": "inactive-star.svg",
      }
    }
  />
  <img
    alt="card background"
    src="rebel-alliance.png"
  />
</button>
`;

exports[`Buttons Favorite Card Button Should be disabled/enabled if active state is true/false 2`] = `
<button
  className="active-fav favorite-card-btn"
  name="favorite-card"
  onClick={[MockFunction]}
  title="Add to Favorites"
  type="button"
>
  <img
    alt="card background"
    src="empire.png"
  />
  <img
    alt="card background"
    src={
      Object {
        "ReactComponent": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "__esModule": true,
        "default": "active-star.svg",
      }
    }
  />
  <img
    alt="card background"
    src="rebel-alliance.png"
  />
</button>
`;

exports[`Buttons ViewFavorites Button: Should be disabled with favorite count set to 0 1`] = `
<button
  className="active view-fav-btn"
  disabled={true}
  name="favorites"
  onClick={[MockFunction]}
  type="button"
>
  No favorites
</button>
`;

exports[`Buttons ViewFavorites Button: Should match snapshot taken 1`] = `
<button
  className="active view-fav-btn"
  disabled={false}
  name="favorites"
  onClick={[MockFunction]}
  type="button"
>
  Favorites
  <span
    className="fav-total"
    id="favorites"
  >
    1
  </span>
</button>
`;
